[Best Practices] Add best practices article and enhance the Scripting Fundamentals (#277)
OfficeDev/office-scripts-docs: docs/develop/scripting-fundamentals.md

[!INCLUDE [Preview note](../includes/preview-note.md)]

## TypeScript: The language of Office Scripts

> Office Scripts are written in [TypeScript](https://www.typescriptlang.org/docs/home.html), which is a superset of [JavaScript](https://developer.mozilla.org/docs/Web/JavaScript). If you are familiar with JavaScript, your knowledge will carry over (much of your code is the same in both languages). We recommend you have some beginner-level programming knowledge before starting your Office Scripts coding journey. The following resources can help you understand the coding-side of Office Scripts.

Reviewer: ```suggestion
Office Scripts are written in [TypeScript](https://www.typescriptlang.org/docs/home.html), which is a superset of [JavaScript](https://developer.mozilla.org/docs/Web/JavaScript). If you are familiar with JavaScript, your knowledge will carry over (much of your code is the same in both languages). We recommend you have some beginner-level programming knowledge before starting your Office Scripts coding journey. The following resources can help you understand the coding side of Office Scripts.
```

[!INCLUDE [Preview note](../includes/coding-basics-references.md)]

## `main` function: The script's starting point

> Each Office Script must contain a `main` function with the `ExcelScript.Workbook` type as its first parameter. When the function is executed, the Excel application invokes this `main` function by providing the workbook as its first parameter. An `ExcelScript.Workbook` should always be the first parameter.

Reviewer: "Avoid execute except to follow the UI. Use run instead. If the UI includes execute, the resulting action is still run." - From the [style guide](https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=33512). 
```suggestion
Each Office Script must contain a `main` function with the `ExcelScript.Workbook` type as its first parameter. When the function runs, the Excel application invokes this `main` function by providing the workbook as its first parameter. An `ExcelScript.Workbook` should always be the first parameter.
```

```TypeScript
function main(workbook: ExcelScript.Workbook) {
  // Your code goes here
}
```

The code inside the `main` function runs when the script is run. `main` can call other functions in your script, but code that's not contained in a function will not run. Scripts cannot invoke or call other Office Scripts.

> You can use [Power Automate](https://flow.microsoft.com) to connect scripts in flows. Add additional arguments to `main` for flows to pass information to your script. Return something from your script to pass information from the script to the flow. How to integrate Office Scripts with Power Automate is covered in detail in [Run Office Scripts with Power Automate](power-automate-integration.md).

Reviewer 2: ```suggestion
[Power Automate](https://flow.microsoft.com) allows you to connect scripts in flows. Add additional arguments to `main` for flows to pass information to your script. Return something from your script to pass information from the script to the flow. How to integrate Office Scripts with Power Automate is covered in detail in [Run Office Scripts with Power Automate](power-automate-integration.md).
```

### Collections and other object relations

> Any child object can be accessed through its parent object. For example, you can read `Worksheets` from the `Workbook` object. There will be a related `get` method on the parent class that (e.g., `Workbook.getWorksheets()` or `Workbook.getWorksheet(name)`). `get` methods that are singular return a single object and require an ID or name for the specific object (such as the name of a worksheet). `get` methods that are plural (such as `getTables`) return the entire object collection as an array. If the collection is empty, you'll get an empty array (`[]`).

Reviewer 2: Also, as a best practice Microsoft style strongly suggests avoiding using parentheses in sentences. It's better to rewrite sentences to avoid using them. If it's important enough to include, it's important enough to include without parentheses. Also, narrator reads them literally so it's very annoying to someone who relies on the narrator! Not just people with low or no vision, lots of customers who are perfectly sighted use Narrator to read topics to avoid strain on the eyes or to do something else while they listen (like cook dinner!)

    // Get the first worksheet.
    let sheet = workbook.getWorksheets()[0];

    // Add a table that uses the data in A1:G10.
    let table = sheet.addTable(
      "A1:G10",
       true /* True because the table has headers. */
    );
    
    // Give the table a name for easy reference in other scripts.
    table.setName("MyTable");
}
```

> [!TIP]
> Most Excel objects have a `setName` method. This gives you an easy way to access Excel objects later in the script or in other scripts for the same workbook.

#### Check if an object exists in the collection

Scripts often need to check if a table or similar object exists before continuing. You can use the names given by scripts or through the Excel UI to identify necessary objects and act accordingly. `get` methods return `undefined` when the requested object is not in the collection.

> The following script requests a table named "MyTable" and uses an `if/else` statement to check if the table was found.

Reviewer: Mozilla formats this as [if...else](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else), like try...catch. 
```suggestion
The following script requests a table named "MyTable" and uses an `if...else` statement to check if the table was found.
```

```TypeScript
function main(workbook: ExcelScript.Workbook) {
  // Get the table named "MyTable".
  let myTable = workbook.getTable("MyTable");

  // If the table is in the workbook, myTable will have a value.
  // Otherwise, the variable will be undefined and go to the else clause.
  if (myTable) {
    let worksheetName = myTable.getWorksheet().getName();
    console.log(`MyTable is on the ${worksheetName} worksheet`);
  } else {
    console.log(`MyTable is not in the workbook.`);
  }
}
```

A common pattern in Office Scripts is to recreate a table, chart, or other object every time the script is run. If you don't need the old data, it's best to delete the old object before creating the new one. This avoids name conflicts or other differences that may have been introduced by other users.

> The following script removes the table named "MyTable", if it is present, then adds a table with the same name.

Reviewer: It's kind of a confusing concept so consider including this extra detail. 
```suggestion
The following script removes the table named "MyTable", if it is present, then adds a new table with the same name.
```
